### ISSUE TITLE
Creating new append blob with append_blob_from_text results in block blob

### ISSUE THREAD

**Issue by Botje** [Azure/azure-storage-python#633]
### Which service(blob, file, queue) does this issue concern?
blob

### Which version of the SDK was used? Please provide the output of `pip freeze`.
2.1.0

pip freeze:
```
asn1crypto==0.24.0
azure-common==1.1.23
azure-functions==1.0.0b5
azure-storage-blob==2.1.0
azure-storage-common==2.1.0
certifi==2019.6.16
cffi==1.12.3
chardet==3.0.4
cryptography==2.7
idna==2.8
pycparser==2.19
python-dateutil==2.8.0
requests==2.22.0
six==1.12.0
urllib3==1.25.3
```

### What problem was encountered?
I want to append to a blob from several writers, but the first write needs to add a header. Therefore, I use the `if_none_match="*"` option to ignore the write if the file already exists.
I tried to create a new append blob with content using `append_blob_from_text` and a non-existent blob name, with the intent of appending further content later.

This code:
```
appendservice.append_blob_from_text(storage_container, 'blub', if_none_match='*', text="HEADER")
```
results in a **block** blob, which cannot be appended to.

### Have you found a mitigation/solution?
No. Calling `create_blob` separately results in a race condition where the blob is created sans header.

Note: for table service, please post the issue here instead: https://github.com/Azure/azure-cosmosdb-python.

**Comment by xiafu-msft** [comment]
Hi @Botje,

Thank for reaching out! Sorry that the document of create_blob maybe is misleading, we will try to improve that later.
Here is How you should use append_blob_from_text, you need to call create_blob first, else you should get Exception saying some thing like the blob doesn't exist. 
https://github.com/Azure/azure-storage-python/blob/690f7c44f6770603adf642926a40151543ad9ee2/tests/blob/test_append_blob.py#L796

Maybe you can call create_blob and then you will be able to get the ETag value for the blob on creation. Use that ETag value to determine if you should append "HEADER" or not.

**Comment by Botje** [comment]
Does the ETag value for blobs depend on anything other than their contents? If so, how can I calculate it off-line?

I want to append from multiple azure functions at the same time, so my strategy would be:

```
create_blob(..., if_none_match="*") # ignore any exceptions
append_to_blob(..., header, if_match=ETAG_FOR_EMPTY_BLOB) # ignore any exceptions
append_to_blob(..., data)
```

**Comment by xiafu-msft** [comment]
Hi @Botje,
ETag is a service provided value. The value will be changed every time you edit the blob.

So after you call create_blob, you will get an ETag value in response, that value is the one you want to use for ETAG_FOR_EMPTY_BLOB

**Comment by Botje** [comment]
Hi @xiafu-msft , thanks for your answers so far.

Unfortunately, this leaves me with an unsolvable race condition. As I mentioned in the issue description, I want to construct an append blob with multiple simultaneous writer processes (well, Azure functions) with the guarantee that there is exactly one header at the beginning of the blob.

To keep it simple, let's say I have two concurrent processes running the following code:

```
try:
    etag, _ = create_blob(container, 'my_blob', if_none_match="*")
    append_blob_from_text(container, 'my_blob', "HEADER", if_match=etag)
except:
    pass
append_blob_from_bytes(container, 'my_blob', b"<other data>")
```

It is possible that two processes start running very close to each other, so:
- Both processes try the `create_blob`. One (call it P1) succeeds, the other (P2) receives an `AzureHttpConflictException` and exits the try-except block. 
- P1 now tries to append the header, while P2 attempts to append its data. If P1 wins the race we have `HEADER<other data>`. If P2 wins the race, we now have a file of the form `<other data>HEADER`. 

Should somehow _both_ processes succeed in the `create_blob` step, we can use the ETag mechanism to ensure only one writes the header. This might still leave a small window where P2 can sneak in the `<other data>` before P1 finishes writing the header, but that might be impossible given the details of the Azure architecture.

I would like to avoid the `<other data>HEADER` situation, which is why I was hoping for an atomic `create_blob_from_*` function or an `append_blob_from_*` function that can creates append blobs.

**Comment by zezha-msft** [comment]
> results in a block blob, which cannot be appended to.

Hi @Botje, to clarify, calling `append_blob_from_text` against a missing blob does not result in a block blob, an error like the following should be returned:
```
azure.common.AzureMissingResourceHttpError: The specified blob does not exist. ErrorCode: BlobNotFound
<?xml version="1.0" encoding="utf-8"?><Error><Code>BlobNotFound</Code><Message>The specified blob does not exist.
RequestId:b82d7f7e-601e-001c-0145-64173c000000
Time:2019-09-05T23:57:34.6637090Z</Message></Error>
```

In addition, you seem to have an application design issue, since the SDK is not responsible for managing concurrency. In order to achieve the behavior that you desire, here are some recommendations:

1. The easiest solution is to use a distributed mutex, which can be implemented with the help of [blob lease](https://docs.microsoft.com/en-us/rest/api/storageservices/lease-blob). An example is [here](https://github.com/Azure/azure-storage-azcopy/blob/dev/tool_distributed_mutex.py).
    - In other words, you could use another blob as the global lock to create blobs, e.g. all the Azure Functions can try to get the lease, the one that gets it can create the append blob and append the header; since the create calls have the condition `if_none_match="*"`, the blob only gets created once.
    - If new instances of Azure Functions get spawn up, you could perhaps check whether the append blob exists already, before attempting to create it.
2. Perhaps you could modify the application to not require such a single blob where every Azure Functions write to, as it adds complexity due to the need for synchronization.
